modules/schema/src/main/java/org/apache/ignite/internal/schema/row/BinaryTuple.java

```diff
@@ -0,0 +1,223 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.schema.row;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.util.BitSet;
+import java.util.UUID;
+import org.apache.ignite.internal.schema.InvalidTypeException;
+
+/**
+ * General interface to describe binary tuples outside of their data layout and column schemas.
+ * Accessor methods may or may not throw {@link InvalidTypeException} depending on the implementation.
+ */
+public interface BinaryTuple {
+    /**
+     * Returns a number of values in the tuple.
+     */
+    int count();
+
+    /**
+     * Checks whether the given column contains a null value.
+     *
+     * @param col Column index.
+     * @return {@code true} if this column contains a null value, {@code false} otherwise.
+     */
+    boolean hasNullValue(int col);
+
```

Comment:
I'd suggest to add another method:

`ByteBuffer getRawData(int col)` -- it will return `null` for NULL and real sub-buffer to the underlying binary field otherwise.

Reply:
So basically you imply that there's information about element size in the internal structure of the tuple. I disagree with this statement, this can only be applied to some specific implementations. And, I don't get why we would need such method in  the first place

Reply:
This interface already requires that there is information on where each element starts. It seems natural that you can also figure out where an element ends given its start and knowledge about tuple internals.

Also for varlen elements the size information must be present in any case.

In principle this method could be defined to return valid data only for varlens while for fixed-size elements it could be allowed to throw.

This method might be needed for:
  1. schema-less copying of individual columns to other tuples (e.g. by SQL-engine when doing projections, etc);
  2. cheap calculation of hash value for a byte field, etc.

Reply:
> schema-less copying of individual columns to other tuples (e.g. by SQL-engine when doing projections, etc);

> cheap calculation of hash value for a VARBINARY element, etc.

Both of these statements imply that data encoding algorithm is known and the same for every implementation, which is simply not true.
Plus, I would want to have `Object[]`-based implementation, for which it would be impossible to implement a method like this.

Reply:
> Plus, I would want to have Object[]-based implementation, for which it would be impossible to implement a method like this.

Then this interface shouldn't be named `BinaryTuple`.

Reply:
Yeah, about that... let's rename it. I can't think of a good name. We already have "org.apache.ignite.table.Tuple", so simply calling it Tuple is a bad idea. SchemalessTuple maybe? I don't know

Reply:
> Then this interface shouldn't be named BinaryTuple.

Yes, I agree